File: bin/adjustOpticalMaps.py
#!/usr/bin/env python
# adjustOpticalMaps.py
# Author: Lee Mendelowitz (lmendelo@umiacs.umd.edu)
# Date: 1/26/2013 
#
# Wrapper around adjustOpticalMaps.R
# Run the R script to adjust optical map fragments
# to remove the oversizing bias from a finished optical map.
# The bias is removed by using loess regression to predict the
# true restriction fragment size from the reported optical map
# restriction fragment size.

import argparse
import subprocess
import os
import sys

parser = argparse.ArgumentParser(description="adjustOpticalMaps: Remove oversizing bias from optical maps.")
parser.add_argument('inputMapFile', help='path to the input optical map file')
parser.add_argument('matchFragFile', help='path to the training data for loess regression')
parser.add_argument('outputMapFile', help='output optical map file')

R_SCRIPT = 'adjustOpticalMaps.R'

def MissingInputFileError(Exception):
    pass

def RFailure(Exception):
    pass


def checkFileExists(fname):
    if not os.path.exists(fname):
        raise MissingInputFileError('Input file does not exist: %s'%fname)

def run(mapFile, matchFragFile, outputMapFile):
    cmd = ['R', '--quiet', '--slave', 
            '--args',
            mapFile,
            matchFragFile,
            outputMapFile,
            ]
    #sys.stdout.write(' '.join(cmd) + '\n')
    #sys.stdout.flush()
    p = subprocess.Popen(cmd, stdin = subprocess.PIPE)
    script = open(R_SCRIPT).read()
    p.stdin.write(script)
    p.stdin.close()
    ret = p.wait()
    if (ret != 0):
        raise RFailure('R Script returned with error code: %i'%ret)

def makeMatchedChunkFile(matchList, outputFile):
    alns = [mc for mr in matchList for mc in mr.alignment if not (mc.isBoundaryChunk or mc.isContigGap)]
    contigChunks = [al.contig.lengthBp() for al in alns]
    opticalChunks = [al.optical.lengthBp() for al in alns]
    fout = open(outputFile, 'w')
    fout.write('contig\toptical\n')
    for c, o in zip(contigChunks, opticalChunks):
        fout.write('%i\t%i\n'%(c,o))
    fout.close()

def main():

    args = parser.parse_args()

    # Check that input files exist
    die = False
    checkFileExists(args.inputMapFile)
    checkFileExists(args.matchFragFile)
    checkFileExists(R_SCRIPT)

    # Run the R Script
    run(args.inputMapFile, args.matchFragFile, args.outputMapFile)


if __name__ == '__main__':
    main()

File: bin/adjustOpticalMaps.R
c# Lee Mendelowitz
# LMendelo@umiacs.umd.edu
# File: adjustOpticalMaps.R
# 
# Apply loess regression to remove bias in reported optical map fragment sizes
###############################################################################

############################################
# Convert a map to a string for output
mapToStr <- function(map)
{
  sout = paste(map$mapId, map$L, map$numFrags, sep="\t")
  fragStr = paste(map$frags, collapse="\t")
  sout = paste(sout, fragStr, sep="\t")
  return(sout)
}

############################################
# Convert a string (line) into a map
strToMap <- function(line)
{
  fields = strsplit(line, "\t")[[1]]
  n = length(fields)
  mapId = fields[1]
  L = fields[2]
  numFrags = fields[3]
  frags = as.numeric(fields[4:n])
  bp = sum(frags)
  stopifnot(length(frags)==numFrags)
  stopifnot(L==bp)
  map = list(mapId=mapId, L=L, numFrags=numFrags,frags=frags)
  class(map) <- "SOMAMap"
  return(map)
}

############################################
# Write the list of maps to the file handle
writeMaps <- function(maps, fname)
{
  handle = file(description=fname, open="w")
  mapStrs = lapply(maps, mapToStr)
  writeMap <- function(myStr) {write(myStr, file = handle, append=TRUE, sep="\n")}
  for(mapStr in mapStrs)
  {
    writeMap(mapStr)
  }
  close(handle)
}

############################################
# Read maps from a file, return a list of maps
readMaps <- function(fname)
{
  lines = readFile(fname)
  maps = lapply(lines, strToMap)
}

############################################
# Read lines from a file
readFile <- function(fname)
{
    lines <- scan(file=fname, what = character(), sep="\n", quiet=TRUE)
    return(lines)
}

############################################
runLoess <- function(alignedChunkFile)
{
  t <- read.table(alignedChunkFile, header=TRUE, sep="\t")
  fit <- loess(t$contig ~ t$optical)
  return(fit)
}

###########################################
# Adjust the map fragments by applying new frags
adjustMap <- function(map, fit)
{
  newMap = map
  newMap$frags <- predict(fit, map$frags)
  # If the fit gave NA, use the original value
  naInd = which(is.na(newMap$frags))
  if(length(naInd) > 0)
  {
    cat(sprintf("\n\nWARNING: Loess adjustment gave NA for these fragments in map %s:\n", map$mapId))
    origVals = map$frags[naInd]
    ind2Str <- function(ind) {s = sprintf('[%d]=%d',ind,map$frags[ind])}
    cat(paste(lapply(naInd, ind2Str)))
    cat ("\nKeeping these fragments as original values.\n\n")
    newMap$frags[naInd] = map$frags[naInd]    
  }
  newMap$frags = round(newMap$frags)
  newMap$L = sum(newMap$frags)
  return(newMap)
}

############################################
run <- function(mapFileIn, alignedChunkFileIn, adjustMapFileOut)
{
  maps <- readMaps(mapFileIn)
  cat(sprintf("Read %d maps from file %s.\n", length(maps), mapFileIn))
  
  fit = runLoess(alignedChunkFileIn)
  cat(sprintf("Ran loess fit using aligned chunk file %s\n", alignedChunkFileIn))
  
  newMaps = lapply(maps, function(map){adjustMap(map,fit)} )
  cat("Done adjusting maps.\n")
  
  writeMaps(newMaps, adjustMapFileOut)
  cat(sprintf("Wrote adjusted maps to output file %s.\n", adjustMapFileOut))
}

############################################
main <- function()
{
  args = parseArgs()
  mapFileInput = args[1]
  matchedChunksFile = args[2]
  mapFileOutput = args[3]
  run(mapFileInput, matchedChunksFile, mapFileOutput)
}

############################################
printUsage <- function()
{
  msg = paste("To run this script from commandline:\n\n",
          "R --quiet --slave --args MAP_FILE_INPUT MATCHED_CHUNKS_FILE MAP_FILE_OUTPUT < adjustOpticalMaps.R\n\n\n",
          "DESCRIPTION:\n",
          "This will adjust the input maps specified in MAP_FILE_INPUT by performing loess\n",
          "regression on the matched chunks in MATCHED_CHUNKS_FILE.\n",
          "The adjusted maps will be written to MAP_FILE_OUTPUT\n")
 cat(msg)  
}

############################################
parseArgs <- function()
{
  args = commandArgs(trailingOnly=TRUE)
  
  if (length(args) < 3)
  {
    cat('ERROR: Too few arguments!\n')
    printUsage()
    q()
  }
  
  if (length(args) > 3)
  {
    error('ERROR: Too many arguments!\n')
    printUsage()
    q()
  }
  
  mapFileInput = args[1]
  matchedChunksFile = args[2]
  mapFileOutput = args[3]
  
  msg = sprintf('MapFileInput: %s\nMatchedChunkInput: %s\nMapFileOutput: %s\n',
                mapFileInput, matchedChunksFile, mapFileOutput)
  cat(msg)
  return(c(mapFileInput, matchedChunksFile, mapFileOutput))
}

############################################n

main()
